lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -489,6 +489,34 @@ class DecomposeAtenMeanOp : public OpRewritePattern<AtenMeanOp> {
 };
 } // namespace
 
+// Decompose `AtenMaxDimOp` into `CustomAtenMaxDimOp` and `AtenArgmaxOp`
+// respectively. `CustomAtenMaxDimOp` is a custom version of `AtenMaxDimOp` that
+// only returns `values` Tensor.
+namespace {
+class DecomposeAtenMaxDimOp : public OpRewritePattern<AtenMaxDimOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenMaxDimOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value input = op.self();
+    Value dim = op.dim();
+    Value keepDim = op.keepdim();
+    Value values = op.values();
+    Value indices = op.indices();
+    BaseTensorType valuesTensorType = values.getType().cast<BaseTensorType>();
+    BaseTensorType indicesTensorType = indices.getType().cast<BaseTensorType>();
+
+    Value maxResult = rewriter.create<CustomAtenMaxDimOp>(loc, valuesTensorType,
+                                                          input, dim, keepDim);
+    Value indicesResult = rewriter.create<AtenArgmaxOp>(loc, indicesTensorType,
+                                                        input, dim, keepDim);
```

Comment:
It's not efficient to decompose to `CustomAtenMaxDimOp` and `AtenArgmaxOp` since the second op calculates both max values and indices in one `linalg.generic` and drop the max value result. The calculation for `CustomAtenMaxDimOp` is redundant. It's probably better to decompose `AtenArgmaxOp` into `aten.max.dim` instead.

Reply:
I didn't get it. The second result of `aten.max.dim` op actually calculates `argmax` and will be required.

Reply:
This might be helpful:

https://github.com/llvm/torch-mlir/blob/cadea678e5e2cfc62a1625af6e50cac34144e710/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L2156-L2159

Its better to use the switch the ops `Argmax` and `MaxDim`. `MaxDim` can be lowered in `TorchToLinalg` using the code that is already there for `Argmax`, and `Argmax` can be moved to the `DecomposeComplexOps` pass and decompose into `MaxDim`

Reply:
I like this, because then we can special-case when the "indices" have no uses and lower to a regular max reduction, because that can be lowered more efficiently.

Reply:
The current implementation of `Argmax` can be mostly reused as it calculates max values and indices already but dropped the max values https://github.com/llvm/mlir-npcomp/blob/d6b6c0268cfd6ead0aa5f2f21ba02bbe8ead3180/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L2237.

Reply:
@silvasean @ramiro050 @cathyzhyi   We can't decompose aten_arg_max to aten_max_dim because the dim can be NoneType in the former(which means reduce along all dims). I have converted the argMaxOp rewrite pattern into a static function which both `argMax` as well as `maxDim` op can call.

Reply:
@cathyzhyi @ramiro050 I have made the required changes, please re-review.